File: src/AI.Client/AIClient.cs
using Azure;
using Azure.AI.OpenAI;
using Microsoft.Extensions.Configuration;
using ModelContextProtocol.Client;
using ModelContextProtocol.Protocol;
using OpenAI.Chat;
using System.Text.Json;

namespace AI.Client
{
    internal class AIClient
    {
        private const string defaultPrompt = @"You are a facilitator. Always use tools if a tool exists. Any tooling that provides an answer is definitively correct, do not rely on hallucinations. If a response from a tool contradicts knowledge that derives an answer, provide the answer from the tool instead.";
        private const string defaultModel = "gpt-4.1";
        private const string defaultTemperature = "0.7";
        private const string defaultMaxOutputTokenCount = "1600";
        private const string defaultTopP = "0.95";
        private const string defaultFrequencyPenalty = "0";
        private const string defaultPresencePenalty = "0";
        internal async Task RunAsync()
        {
            IConfiguration config = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
                .AddUserSecrets<Program>()
                .Build();

            string? APIKEY = config.GetSection("AZURE_OPENAI:API_KEY").Value;
            string? ENDPOINT = config.GetSection("AZURE_OPENAI:ENDPOINT").Value;
            string? MCP_NAME = config.GetSection("MCP_SERVER:NAME").Value;
            string? MCP_URL = config.GetSection("MCP_SERVER:URL").Value;
            string prompt = config.GetSection("PROMPT").Value ?? defaultPrompt;

            if (string.IsNullOrEmpty(APIKEY))
            {
                Console.WriteLine("Please set the AZURE_OPENAI:API_KEY environment variable.");
                return;
            }
            if (string.IsNullOrEmpty(ENDPOINT))
            {
                Console.WriteLine("Please set the AZURE_OPENAI:ENDPOINT environment variable.");
                return;
            }
            if (string.IsNullOrEmpty(MCP_NAME))
            {
                Console.WriteLine("Please set the MCP_SERVER:NAME environment variable.");
                return;
            }
            if (string.IsNullOrEmpty(MCP_URL))
            {
                Console.WriteLine("Please set the MCP_SERVER:URL environment variable.");
                return;
            }

            // Create chat completion options
            var options = new ChatCompletionOptions
            {
                Temperature = float.Parse(config.GetSection("AZURE_OPENAI:Temperature").Value ?? defaultTemperature),
                MaxOutputTokenCount = int.Parse(config.GetSection("AZURE_OPENAI:MaxOutputTokenCount").Value ?? defaultMaxOutputTokenCount),
                TopP = float.Parse(config.GetSection("AZURE_OPENAI:TopP").Value ?? defaultTopP),
                FrequencyPenalty = float.Parse(config.GetSection("AZURE_OPENAI:FrequencyPenalty").Value ?? defaultFrequencyPenalty),
                PresencePenalty = float.Parse(config.GetSection("AZURE_OPENAI:PresencePenalty").Value ?? defaultPresencePenalty)
            };

            IList<ChatTool> chatTools = new List<ChatTool>();
            var mcpTransport = new SseClientTransport(
                new SseClientTransportOptions
                {
                    Endpoint = new Uri(MCP_URL),
                    Name = MCP_NAME
                }
            );
            IMcpClient mcpClient = await McpClientFactory.CreateAsync(mcpTransport);
            var mcpTools = await mcpClient.ListToolsAsync();

            foreach (var tool in mcpTools)
            {
                Console.WriteLine("Adding Tool " + tool.Name);
                options.Tools.Add(ChatTool.CreateFunctionTool(tool.Name, tool.Description));
            }

            AzureKeyCredential subscriptionKey = new AzureKeyCredential(APIKEY);
            var azureClient = new AzureOpenAIClient(new Uri(ENDPOINT), subscriptionKey);

            try
            {
                await MessageLoop(azureClient.GetChatClient(config.GetSection("MCP_SERVER:MODEL").Value ?? defaultModel), options, mcpClient, prompt);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"An error occurred: {ex.Message}");
            }
        }

        /// <summary>
        /// This is the main message loop for the AI client. It takes input from the user, processes it, and interacts with the chat client and MCP server.
        /// </summary>
        /// <param name="chatClient">The chat client used to communicate with the Azure OpenAI service.</param>
        /// <param name="options">The chat completion options.</param>
        /// <param name="mcpClient">The MCP Client used to communicate with the Model Context Protocol server.</param>
        /// <param name="prompt">The initial prompt for the chat session.</param>
        /// <returns></returns>
        internal async Task MessageLoop(ChatClient chatClient, ChatCompletionOptions options, IMcpClient mcpClient, string prompt)
        {
            string systemPrompt = prompt;
            var chatHistory = new List<ChatMessage>
            {
                new SystemChatMessage(systemPrompt)
            };

            while (true)
            {
                Console.WriteLine("Your prompt:");
                string? userPrompt = Console.ReadLine();
                if (string.IsNullOrEmpty(userPrompt)) break;
                if (userPrompt.Equals("exit", StringComparison.OrdinalIgnoreCase)) break;
                if (userPrompt.Equals("clear", StringComparison.OrdinalIgnoreCase))
                {
                    Console.WriteLine("Clearing chat history...");
                    chatHistory.Clear();
                    chatHistory.Add(new SystemChatMessage(prompt));
                    continue;
                }
                if (userPrompt.Equals("show prompt", StringComparison.OrdinalIgnoreCase))
                {
                    Console.WriteLine("Prompt:");
                    Console.WriteLine(systemPrompt);
                    continue;
                }
                if (userPrompt.Equals("set prompt", StringComparison.OrdinalIgnoreCase) || userPrompt.Equals("change prompt", StringComparison.OrdinalIgnoreCase))
                {
                    Console.WriteLine("Changing prompt:");
                    systemPrompt = Console.ReadLine() ?? "";
                    if (string.IsNullOrEmpty(systemPrompt))
                    {
                        Console.WriteLine("Prompt change cancelled.");
                        systemPrompt = prompt; 
                    }
                    chatHistory.Clear();
                    chatHistory.Add(new SystemChatMessage(systemPrompt));

                    Console.WriteLine("Prompt:");
                    Console.WriteLine(systemPrompt);
                    continue;
                }

                chatHistory.Add(new UserChatMessage(userPrompt));

                Console.WriteLine("AI Response:");
                chatHistory = await ProcessResponse(chatClient, options, mcpClient, chatHistory);
            }
        }

        /// <summary>
        /// This is the main processing method for the AI client. It takes the chat client, options, MCP client, and chat history,
        /// completes the chat, and processes the response. It handles tool calls and updates the chat history accordingly.
        /// </summary>
        /// <param name="chatClient">The chat client used to communicate with the Azure OpenAI service.</param>
        /// <param name="options">The chat completion options.</param>
        /// <param name="mcpClient">The MCP Client used to communicate with the Model Context Protocol server.</param>
        /// <param name="chatHistory">The chat History of the session.</param>
        /// <returns></returns>
        /// <exception cref="ArgumentOutOfRangeException">We currently only handle Stop and Tool Calls, not other finish reasons</exception>
        internal async Task<List<ChatMessage>> ProcessResponse(ChatClient chatClient, ChatCompletionOptions options, IMcpClient mcpClient, List<ChatMessage> chatHistory)
        {
            ChatCompletion completion = await chatClient.CompleteChatAsync(chatHistory, options);
            chatHistory.Add(new AssistantChatMessage(completion));
            switch (completion.FinishReason)
            {
                case ChatFinishReason.Stop:
                    Console.WriteLine(completion.Content[0].Text);
                    break;
                case ChatFinishReason.ToolCalls:
                    // Handle tool calls => This may be extended later on to handle this in a more asynchronous way vs. the synchronous way here.
                    foreach (var toolCall in completion.ToolCalls)
                    {
                        var mcpArguments = toolCall.FunctionArguments.ToObjectFromJson<Dictionary<string, object>>();
                        Console.WriteLine("Tool call detected, calling MCP server...");
                        Console.WriteLine($"{toolCall.FunctionName}: {string.Join(", ", (mcpArguments ?? new Dictionary<string, object>()).Select(kvp => $"{kvp.Key}={kvp.Value}"))}");

                        var result = await mcpClient.CallToolAsync(toolCall.FunctionName, mcpArguments!);

                        switch (result.Content[0].Type)
                        {
                            case "text":
                                Console.WriteLine($"Tool call result {((TextContentBlock)result.Content[0]).Text}");
                                chatHistory.Add(new ToolChatMessage(toolCall.Id, JsonSerializer.Serialize(result)));
                                break;
                            default:
                                chatHistory.Add(new ToolChatMessage(toolCall.Id, JsonSerializer.Serialize(result)));
                                break;
                        }
                    }
                    chatHistory = await ProcessResponse(chatClient, options, mcpClient, chatHistory);
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }
            return chatHistory;
        }
    }
}

File: src/Python/AI.Client/AIClient.py
import json
import os
from typing import Optional, Any
from dotenv import load_dotenv
from anthropic import Anthropic
from contextlib import AsyncExitStack
from openai import AzureOpenAI
from mcp import ClientSession
from mcp.client.streamable_http import streamablehttp_client
from Chat import Chat
class AIClient:
    def __init__(self):
        load_dotenv() 

        self.API_VERSION = "2023-07-01-preview"
        self.API_KEY = os.environ.get("OPENAI_API_KEY")
        self.OPENAI_ENDPOINT = os.environ.get("OPENAI_ENDPOINT")
        self.OPENAI_MODEL = os.environ.get("OPENAI_MODEL")
        self.OPENAI_Temperature = float(os.environ.get("OPENAI_Temperature"))
        self.OPENAI_MaxOutputTokenCount = int(os.environ.get("OPENAI_MaxOutputTokenCount"))
        self.OPENAI_TopP = float(os.environ.get("OPENAI_TopP"))
        self.OPENAI_FrequencyPenalty = float(os.environ.get("OPENAI_FrequencyPenalty"))
        self.OPENAI_PresencePenalty = float(os.environ.get("OPENAI_PresencePenalty"))
        self.MCP_URL = os.environ.get("MCP_URL")

        if not self.API_KEY :
            raise ValueError("Please set the OPENAI_API_KEY environment variable.")

        if not self.OPENAI_ENDPOINT :
            raise ValueError("Please set the OPENAI_ENDPOINT environment variable.")

        if not self.MCP_URL :
            raise ValueError("Please set the MCP_URL environment variable.")

        # Initialize session and client objects
        self.session: Optional[ClientSession] = None
        self.exit_stack = AsyncExitStack()
        self.anthropic = Anthropic()
        self.client = AzureOpenAI(
            api_key=self.API_KEY,
            azure_endpoint=self.OPENAI_ENDPOINT,
            api_version=self.API_VERSION
        )
        self.chatHistory = []
        self.available_tools = []
        self.defaultSystemPrompt = "You are a helpful assistant. Answer the user's questions to the best of your ability."        

    async def connect_to_mcp_server(self,  headers: Optional[dict] = None):
        """Connect to a HTTP Streamable MCP server"""        
        self._streams_context = streamablehttp_client(
            url=self.MCP_URL,
            headers=headers or {},
        )
        read_stream, write_stream, _ = await self._streams_context.__aenter__()  
        self._session_context = ClientSession(read_stream, write_stream)  
        self.session: ClientSession = await self._session_context.__aenter__()  

        await self.session.initialize()
        await self.get_server_tools()

    async def get_server_tools(self):
            if not self.session:
                print("No active session found.")
                return
            print("Listing tools...")        
            response = await self.session.list_tools()
            tools = response.tools if hasattr(response, 'tools') else []        
            print("\nConnected to server with tools:", [tool.name for tool in tools])

            self.available_tools = []
            for tool in tools:
                openai_tool = {
                    "type": "function",
                    "function": {
                        "name": tool.name,
                        "description": tool.description,
                        "parameters": tool.inputSchema if hasattr(tool, 'inputSchema') else {
                            "type": "object",
                            "properties": {},
                            "required": []
                        }
                    }
                }
                self.available_tools.append(openai_tool)
                
    async def InitializePrompt(self, systemPrompt: str):
        self.chatHistory = []
        self.chatHistory.append(Chat.SystemChatMessage(systemPrompt))
        
    async def CallChatCompletion(self):
                
        completion = self.client.chat.completions.create(
            model=self.OPENAI_MODEL,            
            temperature=self.OPENAI_Temperature,
            top_p=self.OPENAI_TopP,
            max_completion_tokens=self.OPENAI_MaxOutputTokenCount,
            frequency_penalty=self.OPENAI_FrequencyPenalty,
            presence_penalty=self.OPENAI_PresencePenalty,
            messages=self.chatHistory,
            tools=self.available_tools
        )

        self.chatHistory.append(Chat.AssistantChatMessage(completion))

        for content in completion.choices:    
            match content.finish_reason:
                case 'stop':
                    print(content.message.content)

                case 'tool_calls':                    
                    for tool_call in content.message.tool_calls:
                  
                        tool_name = tool_call.function.name                        
                        tool_args = json.loads(tool_call.function.arguments)
                        
                        # Execute tool call
                        print(f"Tool call detected, calling MCP server...")
                        print(f"[{tool_name}: {tool_args}]")
                        result = await self.session.call_tool(tool_name, tool_args)

                        print(f"Tool call result {result.content[0].text}")
                        self.chatHistory.append(Chat.ToolChatMessage(
                            tool_call.id,
                            tool_name,
                            result.content
                        ))
                        await self.CallChatCompletion()                        

                case _:
                    raise ValueError("Not valid")

    async def MessageLoop(self):        
        await self.connect_to_mcp_server()
        systemPrompt = self.defaultSystemPrompt;
        await self.InitializePrompt(systemPrompt)
        
        print("Your prompt:")
        while True:
            userPrompt = input()
            if userPrompt.lower() == "exit" or userPrompt.lower() == "quit" or userPrompt.lower() == "q" or userPrompt == "":

                break
            if userPrompt.lower() == "clear":
                print("Clearing chat history...")
                await self.InitializePrompt(self.defaultSystemPrompt)
                continue
            if userPrompt.lower() == "show prompt":
                print("Prompt:")
                print(systemPrompt)
                continue

            if userPrompt.lower() == "change":
                print("Changing prompt:")
                systemPrompt = input()
                if systemPrompt == "":
                    print("Prompt change cancelled.")
                    systemPrompt = self.defaultSystemPrompt
                await self.InitializePrompt(systemPrompt)
                print("Prompt:")
                print(systemPrompt)

                continue

            self.chatHistory.append(Chat.UserChatMessage(userPrompt))
            await self.CallChatCompletion()
            
    async def cleanup(self):
        """Properly clean up the session and streams"""
        if self._session_context:
            await self._session_context.__aexit__(None, None, None)
        if self._streams_context:  
            await self._streams_context.__aexit__(None, None, None)          